book/styles.md

```diff
@@ -881,17 +1017,11 @@ priorities.[^lexicographic]
     in the sequence will work fine as long as no one strings more than
     16 selectors together.
 
-*Descendant Selectors*: When multiple selectors are separated with
-spaces, like `ul b`, that selects all `<b>` elements with a `<ul>`
-ancestor. Implement descendent selectors; scoring for descendent
-selectors works just like for combination selectors. Make sure that
-something like `section .warning` selects warnings inside sections,
-while `section.warning` selects warnings that *are* sections.
-
-*Ancestor Selectors*: an ancestor selector is the inverse of a descendant
-selector - it styles an ancestor according to the presence of a descendant.
-This feature is one of the benefits provided by the
-[`:has` syntax](https://drafts.csswg.org/selectors-4/#relational). However, you
-will find that `:has` is not implemented in any real browser as yet. Can you
-guess why? Hint: try to implement ancestor selectors and analyze the speed of
-your algorithm.
+*Ancestor Selectors*: an ancestor selector is the inverse of a
+descendant selector - it styles an ancestor according to the presence
+of a descendant. This feature is one of the benefits provided by the
+[`:has` syntax](https://drafts.csswg.org/selectors-4/#relational). Try
+to implement ancestor selectors. As I write this, no browser has
+actually implemented `:has`; why? Hint: analyze the asymptotic speed
+of your implementation. There is a clever implementation that is
+*O(1)* amortized per element---can you find it?
```

Comment:
What is the solution to this you have in mind?

Reply:
Suppose you have a selector, `S1:has(S2)`, where we assume `S2` doesn't have a `:has` inside. Then as a pre-pass, match each element against `S2` and store that as a bit on the element. As a second pass over all the elements, store a second bit for whether any descendant matches `S2`, computed by propagating the first bit upwards. That second bit now represents `:has(S2)` for all elements, so you can match `S1:has(S2)` using it.

That doesn't work for some reason? Perhaps I should have thought more about it.

Reply:
Yes it does work.

This is a good learning moment for the student, as to why even this is not good enough. I suggest adding a sentence saying that in real browsers, even O(1) amortized time is not good enough - iterating twice over the tree is too much overhead.

Literally today we are discussing this:

https://github.com/mozilla/standards-positions/issues/528#issuecomment-842148898